### ISSUE TITLE
preprocessed headers lose their guard

### ISSUE THREAD

**Issue by informatimago** [ThrowTheSwitch/Ceedling#806]
The following mini-project demonstrates three problems:

[ceedling-bug-guard.tgz](https://github.com/ThrowTheSwitch/Ceedling/files/12370746/ceedling-bug-guard.tgz)

To test with my patches:
```
tar zxvf ceedling-bug-guard.tgz
ceedling-bug-guard/get-dependencies informatimago
cd ceedling-bug-guard
make
make tests
```

To test with upstream master:
```
tar zxvf ceedling-bug-guard.tgz
ceedling-bug-guard/get-dependencies upstream
cd ceedling-bug-guard
make
make tests
```

`make all` compiles and runs the program successfully, which shows that the headers, `-I` and `#include`s are consistent.
```
16:39[pjb@despina org.xquartz:0 ceedling-bug-guard 14Gi]$ make clean all
rm -rf build
mkdir -p build
mkdir -p build/src/sub/
clang -Isrc -H -c src/sub/foo.c -o build/src/sub/foo.o
. src/sub/foo.h
mkdir -p build/src/add/
clang -Isrc -H -c src/add/foo.c -o build/src/add/foo.o
. src/add/foo.h
mkdir -p build/src/main/
clang -Isrc -H -c src/main/main.c -o build/src/main/main.o
. src/bar.h
.. src/foo.h
... src/add/foo.h
... src/sub/foo.h
. /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/stdio.h
.. /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/_stdio.h
... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/cdefs.h
.... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_symbol_aliasing.h
.... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_posix_availability.h
... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/Availability.h
.... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/AvailabilityVersions.h
.... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/AvailabilityInternal.h
... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/_types.h
.... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types.h
..... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/machine/_types.h
...... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/i386/_types.h
..... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_pthread/_pthread_types.h
... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_va_list.h
.... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/machine/types.h
..... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/i386/types.h
...... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_int8_t.h
...... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_int16_t.h
...... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_int32_t.h
...... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_int64_t.h
...... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_u_int8_t.h
...... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_u_int16_t.h
...... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_u_int32_t.h
...... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_u_int64_t.h
...... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_intptr_t.h
....... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/machine/types.h
...... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_uintptr_t.h
... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_size_t.h
... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_null.h
... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/stdio.h
.. /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/_ctermid.h
.. /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_off_t.h
.. /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/sys/_types/_ssize_t.h
.. /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/secure/_stdio.h
... /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/secure/_common.h
clang -Isrc -H -o build/pgm build/src/sub/foo.o build/src/add/foo.o build/src/main/main.o
build/pgm
a.val=3
b.val=-1

```

Unfortunately running `Ceedling` fails.

With upstream master:
```
Compiling test_foo.c...
Command: {:line=>"gcc -I\"/Users/pjb/works/harman/upstream/Ceedling/vendor/unity/src\" -I\"/Users/pjb/works/harman/upstream/Ceedling/vendor/c_exception/lib\" -I\"/Users/pjb/works/harman/upstream/Ceedling/vendor/cmock/src\" -I\"build/test/mocks\" -I\"test\" -I\"test/add\" -I\"test/sub\" -I\".\" -I\"src\" -I\".\" -I\"src\" -I\"test\" -DTEST -DUNITY_EXCLUDE_MATH_H -DUNITY_EXCLUDE_STDINT_H -DUNITY_EXCLUDE_LIMITS_H -DUNITY_EXCLUDE_TIME_H -DUNITY_EXCLUDE_STDLIB_H -DUNITY_SUPPORT_64 -DCMOCK_MEM_SIZE=131072 -DGNU_COMPILER -g -c \"test/test_foo.c\" -o \"build/test/out/c/test_foo.o\" -MMD -MF \"build/test/dependencies/test_foo.d\"", :options=>{:stderr_redirect=>:none, :background_exec=>:none}}
Verbose: exec(): gcc -I"/Users/pjb/works/harman/upstream/Ceedling/vendor/unity/src" -I"/Users/pjb/works/harman/upstream/Ceedling/vendor/c_exception/lib" -I"/Users/pjb/works/harman/upstream/Ceedling/vendor/cmock/src" -I"build/test/mocks" -I"test" -I"test/add" -I"test/sub" -I"." -I"src" -I"." -I"src" -I"test" -DTEST -DUNITY_EXCLUDE_MATH_H -DUNITY_EXCLUDE_STDINT_H -DUNITY_EXCLUDE_LIMITS_H -DUNITY_EXCLUDE_TIME_H -DUNITY_EXCLUDE_STDLIB_H -DUNITY_SUPPORT_64 -DCMOCK_MEM_SIZE=131072 -DGNU_COMPILER -g -c "test/test_foo.c" -o "build/test/out/c/test_foo.o" -MMD -MF "build/test/dependencies/test_foo.d"
In file included from test/test_foo.c:2:
src/bar.h:5:5: error: unknown type name 'add'
    5 |     add a;
      |     ^~~
src/bar.h:6:5: error: unknown type name 'sub'
    6 |     sub s;
      |     ^~~
test/test_foo.c: In function 'test_operate':
test/test_foo.c:8:5: error: unknown type name 'add'
    8 |     add a;
      |     ^~~
test/test_foo.c:9:5: error: unknown type name 'sub'
    9 |     sub b;
      |     ^~~
test/test_foo.c:11:5: warning: implicit declaration of function 'sub_operate_ExpectAndReturn'; did you mean 'add_operate_ExpectAndReturn'? []8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wimplicit-function-declaration-Wimplicit-function-declaration]8;;]
   11 |     sub_operate_ExpectAndReturn(1, 2, -1);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |     add_operate_ExpectAndReturn
test/test_foo.c:12:5: warning: implicit declaration of function 'operate' []8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wimplicit-function-declaration-Wimplicit-function-declaration]8;;]
   12 |     operate(a, 1, 2);
      |     ^~~~~~~
ERROR: Shell command failed.
```

With informatimago branches:

```
ERROR: Shell command failed.
> Shell executed command:
'gcc -I"/Users/pjb/works/harman/informatimago/Ceedling/vendor/unity/src" -I"/Users/pjb/works/harman/informatimago/Ceedling/vendor/c_exception/lib" -I"/Users/pjb/works/harman/informatimago/Ceedling/vendor/cmock/src" -I"build/test/mocks" -I"test" -I"test/add" -I"test/sub" -I"." -I"src" -I"." -I"test" -I"src" -DTEST -DUNITY_EXCLUDE_MATH_H -DUNITY_EXCLUDE_STDINT_H -DUNITY_EXCLUDE_LIMITS_H -DUNITY_EXCLUDE_TIME_H -DUNITY_EXCLUDE_STDLIB_H -DUNITY_SUPPORT_64 -DCMOCK_MEM_SIZE=131072 -DGNU_COMPILER -g -c "test/test_foo.c" -o "build/test/out/c/test_foo.o" -MMD -MF "build/test/dependencies/test_foo.d" 2>&1'
> Produced output:
In file included from test/test_foo.c:3:
build/test/mocks/foo.h:28:27: error: conflicting types for 'add'; have 'struct <anonymous>'
   28 | typedef struct {int val;} add;
      |                           ^~~
In file included from src/bar.h:3,
                 from test/test_foo.c:2:
build/test/mocks/foo.h:28:27: note: previous declaration of 'add' with type 'add'
   28 | typedef struct {int val;} add;
      |                           ^~~
build/test/mocks/foo.h:30:27: error: conflicting types for 'sub'; have 'struct <anonymous>'
   30 | typedef struct {int val;} sub;
      |                           ^~~
build/test/mocks/foo.h:30:27: note: previous declaration of 'sub' with type 'sub'
   30 | typedef struct {int val;} sub;
      |                           ^~~
In file included from build/test/mocks/mock_foo.h:7,
                 from test/test_foo.c:4:
build/test/mocks/foo.h:28:27: error: conflicting types for 'add'; have 'struct <anonymous>'
   28 | typedef struct {int val;} add;
      |                           ^~~
build/test/mocks/foo.h:28:27: note: previous declaration of 'add' with type 'add'
   28 | typedef struct {int val;} add;
      |                           ^~~
build/test/mocks/foo.h:30:27: error: conflicting types for 'sub'; have 'struct <anonymous>'
   30 | typedef struct {int val;} sub;
      |                           ^~~
build/test/mocks/foo.h:30:27: note: previous declaration of 'sub' with type 'sub'
   30 | typedef struct {int val;} sub;
      |                           ^~~
test/test_foo.c: In function 'test_operate':
test/test_foo.c:12:5: warning: implicit declaration of function 'operate' [-Wimplicit-function-declaration]
   12 |     operate(a, 1, 2);
      |     ^~~~~~~
> And exited with status: [1].
```

1- `build/test/mocks/foo.h` contains only the pre-processed `src/add/foo.h`
This is the classic bad include handling of Ceedling cf. https://github.com/ThrowTheSwitch/Ceedling/issues/796

2- `build/test/mocks/mock_foo.h` contains only the declarations from `src/foo.h` and included files, but not the macros (notably not the important `operate` macro that expands to a `_Generic()` call.
I would suppose that the `#Include "foo.h"` in `mock_foo.h` was intended to include the mocked file to get macros? But if that was the case we'd get duplicate declarations with non-macro stuff.

3- the last problem that we see with the informatimago branches, but we could also see it with a different project on the master branch, is that as `#define` are removed in pre-processed files, the all important guard #ifndef AAA_BBB__CCC_H / #define AAA_BBB_CCC_H / #endif are removed too in the pre-processed files.  Therefore the contents of `foo.h` are duplicated, and when the test source includes a non-mocked header that includes a header that has been pre-processed, the guard is useless, and declarations are duplicated.

(Apparently, Ceedling doesn't take into account the `:environment: - CC = clang` so it uses gcc, and we don't benefit the `-H` option to dump the includes, but that's what's occurring.

In short it's a mess, and I wonder on what principle we should move to correct the situation.

I've not thought it out enough, but it seems to me that the pre-processed files should be used as input to the cmock process, but should not be included and compiled, only the original files, plus the generated mock_*.[hc] files should be compiled.

Also the paths in the `#include` should not be modified, to ensure that the order of search directories is maintained, and the right headers are always included.

I wonder what other conceptual rules we must have concerning this problem which is at the core of Ceedling.

**Comment by mkarlesky** [comment]
There's a lot here!

Strictly speaking, `#include` guards should not be required as preprocessed files are only intended to be input to mock generation and test runner generation. That preprocessed files end up in search paths is a bug. It was never originally intended to work this way. Over many years various PRs have had unintended consequences that do not break Ceedling but inadvertently modify its behavior because Ruby and especially Rake are quite flexible. Many complicated features and interactions have crept in.

At any rate, in the branch for version 0.32, issue #796 has been fixed. Further, though they should never be needed, `#include` guards (with useful comments) are now generated in preprocessed .h files. And, search paths for test suite builds can now be far better controlled—down to the level of individual test executable builds. We're not there yet, but eventually search paths will also be used to differentiate files of the same name in different places in a directory tree (e.g. needing to mock two different `foo.h` files).

Preprocessing is undergoing a major rework in 0.32. The ordering of steps is now correct again, and `#include` extraction is much more robust and accurate. We have work yet to do on preprocessing test files themselves. In both cases, without a dedicated C code preprocessor that provides all the information we need, preprocessing just can't be 100% correct. But, as we are able, we're adding other features to help fill those gaps in a friendly and reliable way.
